### ISSUE TITLE
[FR] Fancy plotting of the ticks in log scale

### ISSUE THREAD

**Issue by getzze** [JuliaPlots/Plots.jl#3318]
I have been annoyed by the ticks in logscale for some time now, so I finally wrote a `fancylogscale!` function to improve it and I wanted to share it with you all in case somebody needs it.
Tested with PyPlot and GR.

The function:

```julia
"""
    get_tickslogscale(lims; skiplog=false)
Return a tuple (ticks, ticklabels) for the axis limit `lims`
where multiples of 10 are major ticks with label and minor ticks have no label
skiplog argument should be set to true if `lims` is already in log scale.
"""
function get_tickslogscale(lims::Tuple{T, T}; skiplog::Bool=false) where {T<:AbstractFloat}
    mags = if skiplog
        # if the limits are already in log scale
        floor.(lims)
    else
        floor.(log10.(lims))
    end
    rlims = if skiplog; 10 .^(lims) else lims end

    total_tickvalues = []
    total_ticknames = []

    rgs = range(mags..., step=1)
    for (i, m) in enumerate(rgs)
        if m >= 0
            tickvalues = range(Int(10^m), Int(10^(m+1)); step=Int(10^m))
            ticknames  = vcat([string(round(Int, 10^(m)))],
                              ["" for i in 2:9],
                              [string(round(Int, 10^(m+1)))])
        else
            tickvalues = range(10^m, 10^(m+1); step=10^m)
            ticknames  = vcat([string(10^(m))], ["" for i in 2:9], [string(10^(m+1))])
        end

        if i==1
            # lower bound
            indexlb = findlast(x->x<rlims[1], tickvalues)
            if isnothing(indexlb); indexlb=1 end
        else
            indexlb = 1
        end
        if i==length(rgs)
            # higher bound
            indexhb = findfirst(x->x>rlims[2], tickvalues)
            if isnothing(indexhb); indexhb=10 end
        else
            # do not take the last index if not the last magnitude
            indexhb = 9
        end

        total_tickvalues = vcat(total_tickvalues, tickvalues[indexlb:indexhb])
        total_ticknames = vcat(total_ticknames, ticknames[indexlb:indexhb])
    end
    return (total_tickvalues, total_ticknames)
end

"""
    fancylogscale!(p; forcex=false, forcey=false)
Transform the ticks to log scale for the axis with scale=:log10.
forcex and forcey can be set to true to force the transformation
if the variable is already expressed in log10 units.
"""
function fancylogscale!(p::Plots.Subplot; forcex::Bool=false, forcey::Bool=false)
    kwargs = Dict()
    for (ax, force, lims) in zip((:x, :y), (forcex, forcey), (xlims, ylims))
        axis = Symbol("$(ax)axis")
        ticks = Symbol("$(ax)ticks")

        if force || p.attr[axis][:scale] == :log10
            # Get limits of the plot and convert to Float
            ls = float.(lims(p))
            ts = if force
                (vals, labs) = get_tickslogscale(ls; skiplog=true)
                (log10.(vals), labs)
            else
                get_tickslogscale(ls)
            end
            kwargs[ticks] = ts
        end
    end

    if length(kwargs) > 0
        plot!(p; kwargs...)
    end
    p
end
fancylogscale!(p::Plots.Plot; kwargs...) = (fancylogscale!(p.subplots[1]; kwargs...); return p)
fancylogscale!(; kwargs...) = fancylogscale!(plot!(); kwargs...)
```
Test code:

```julia
using Plots
x = range(1e-2, 1e3, length=100)
plot(x, x.^2; xscale=:log10, yscale=:log10)
```
![plot_gr_no_fancylogscale](https://user-images.githubusercontent.com/6518000/108886193-9661c700-7600-11eb-9c2e-0a399b6e9591.png)

```julia
fancylogscale!()
```
![plot_gr_fancylogscale](https://user-images.githubusercontent.com/6518000/108886203-9b267b00-7600-11eb-9769-d2dcda479d8d.png)

Forcing the transform in logscale if you plot `log10.(y)` (for instance if you use StatsPlots, some functions do not work with `scale=:log10`):

```julia
y = log10.(x.^2)
plot(x, y; xscale=:log10)
```
![plot_pyplot_no_fancylogscale_forcey](https://user-images.githubusercontent.com/6518000/108886744-41728080-7601-11eb-8cfa-0c4043be47f8.png)

```julia
fancylogscale!(; forcey=true)
```
![plot_pyplot_fancylogscale_forcey](https://user-images.githubusercontent.com/6518000/108886762-45060780-7601-11eb-8083-6b755e78b0fb.png)

**Comment by isentropic** [comment]
I feel this should be the way it works by default

**Comment by gertvv** [comment]
Thanks, I agree this is really useful.

I have one suggestion though: "minor ticks have no label" - this is fine in general but not when the plot is very "zoomed in". It'd be nice if this was an option, or if minor tick labels were on by default when there are very few ticks.

I tweaked the code by setting `ticknames  = [string(v) for v in tickvalues]` to produce this plot, otherwise it would have had just the "1" label, which I found disorienting:

![image](https://user-images.githubusercontent.com/80785/117966568-57503180-b324-11eb-8553-da00ac288c90.png)
